### PR TITLE
Close leak of progress of invisible units and courses

### LINKED ISSUE #735: 🐛 BUG: Progress reports for hidden course is visible in profile
On the profile page of the live system I can see that I'm a member of the course _MPSE Demo-Kurs_.
However this course isn't visible on the dashboard and if I click the link on the profile page I get a page not found error.
screenshot-2018-5-11 user steffen geli

It should be deciced if a student still can access or see a hidden course if he/she is already a member of this course. In this case the course needs to be visible and accessible from the dashboard too.

If a hidden course shouldn't be accessible by any student, even if they are a member of this course, then the hidden course should be hidden on the profile page too.

## Estimate:
4h

### PR DESCRIPTION
**Pull request by dboschm**
## Description:
Fix Mongo Population in ReportController, so that a user will not get the progress of invisible units or inactive courses

Closes #735  

## Improvements
- Progress of invisible courses is no longer leaked to user

## Known Issues
- ReportController needs a refactoring (maybe extraxt some functions to a dedicated service)
- inconsitent stats can still appear in Teacher statistics view (due to teachers can see hidden units)

### CODE REVIEW

File: api/src/controllers/ReportController.ts [hunk 1]
```diff
@@ -552,14 +552,20 @@ export class ReportController {
...
         name: 1
       })
       .populate({
         path: 'lectures',
         populate: {
-          path: 'units'
+          path: 'units',
+          match: {
+            $and: [
+              {$or: [{visible: undefined}, {visible: true}]}, // TODO For some reason we apparently can have undefined visiblity.
```
Comment (danielkesselberg) [inline api/src/controllers/ReportController.ts]: https://github.com/geli-lms/geli/pull/890 i hope we remove the undefined case. what do you think?

**Comment by dboschm** [comment]
@torss I just reproduced your issue. I'll try to fix this. :(
grafik

**Comment by aeimer** [comment]
Sorry, I have no clue on this one...

File: api/src/controllers/ReportController.ts [hunk 2]
```diff
@@ -567,25 +567,26 @@ export class ReportController {
...
-    const [courses, userProgressData] = await Promise.all([coursesPromise, progressPromise]);
+    const courses = await coursesPromise;
     const courseObjects: any = courses.map((course: ICourseModel) => <ICourse>course.toObject());
-    const courseObjectsWithProgress = courseObjects
-      .map(this.filterUnits)
-      .map(({courseObj, progressableUnitCount}: any) => {
-        const progressStats = this.calculateProgress(userProgressData, progressableUnitCount, courseObj);
+    const aggregatedProgressPromise = courseObjects
+      .map((courseObj: ICourse) => { return this.countUnitsAndRemoveEmptyLectures(courseObj, currentUser); })
+      .map(async ({courseObj, progressableUnitCount, invisibleUnits}: any) => {
```
Comment (danielkesselberg) [inline api/src/controllers/ReportController.ts]: Nice one! I guess collection invisibleUnits in the first place and exclude them from query should do the job.

**Comment by danielkesselberg** [comment]
Good Work @dboschm @grosspersky @torss

**Pull request merged**